File: web_sdk/src/errors/sdk-error.ts
// src/errors/sdk-error.ts

/**
 * All possible SDK error codes.
 */
export enum SdkErrorCode {
  NETWORK_ERROR = 'NETWORK_ERROR',
  AUTH_ERROR = 'AUTH_ERROR',
  DECRYPTION_ERROR = 'DECRYPTION_ERROR',
  KEY_EXCHANGE_ERROR = 'KEY_EXCHANGE_ERROR',
  STORAGE_ERROR = 'STORAGE_ERROR',
  SESSION_NOT_FOUND = 'SESSION_NOT_FOUND',
  FILE_TOO_LARGE = 'FILE_TOO_LARGE',
  DEVICE_LIMIT_REACHED = 'DEVICE_LIMIT_REACHED',
  INVALID_SIGNATURE = 'INVALID_SIGNATURE',
  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
}

/**
 * All SDK errors are instances of SdkError, providing a consistent typed surface.
 * @example
 * try {
 *   await client.openConversation('user-123');
 * } catch (err) {
 *   if (err instanceof SdkError && err.code === SdkErrorCode.AUTH_ERROR) {
 *     // Re-authenticate
 *   }
 * }
 */
export class SdkError extends Error {
  readonly code: SdkErrorCode;
  readonly statusCode?: number;
  readonly cause?: unknown;

  constructor(
    code: SdkErrorCode,
    message: string,
    options?: { cause?: unknown; statusCode?: number },
  ) {
    super(message);
    this.name = 'SdkError';
    this.code = code;
    this.cause = options?.cause;
    if (options?.statusCode !== undefined) {
      // Use Object.defineProperty to bypass readonly+exactOptionalPropertyTypes
      Object.defineProperty(this, 'statusCode', { value: options.statusCode, writable: false, enumerable: true, configurable: false });
    }
    // Ensure correct prototype chain in ES5 transpile scenarios
    Object.setPrototypeOf(this, SdkError.prototype);
  }

  /**
   * Map a server API error response to a typed SdkError.
   */
  static fromApiResponse(
    status: number,
    body: { error_code?: string; message?: string },
  ): SdkError {
    const code = SERVER_ERROR_CODE_MAP[body.error_code ?? ''] ?? SdkErrorCode.UNKNOWN_ERROR;
    const message = body.message ?? `HTTP ${status}`;
    return new SdkError(code, message, { statusCode: status });
  }

  static network(message: string, cause?: unknown): SdkError {
    return new SdkError(SdkErrorCode.NETWORK_ERROR, message, { cause });
  }

  static auth(message: string): SdkError {
    return new SdkError(SdkErrorCode.AUTH_ERROR, message);
  }

  static decryption(message: string, cause?: unknown): SdkError {
    return new SdkError(SdkErrorCode.DECRYPTION_ERROR, message, { cause });
  }

  static storage(message: string, cause?: unknown): SdkError {
    return new SdkError(SdkErrorCode.STORAGE_ERROR, message, { cause });
  }

  static sessionNotFound(userId: string, deviceId: string): SdkError {
    return new SdkError(
      SdkErrorCode.SESSION_NOT_FOUND,
      `No session found for user ${userId} device ${deviceId}`,
    );
  }
}

const SERVER_ERROR_CODE_MAP: Record<string, SdkErrorCode> = {
  unauthorized: SdkErrorCode.AUTH_ERROR,
  forbidden: SdkErrorCode.AUTH_ERROR,
  unknown_tenant: SdkErrorCode.AUTH_ERROR,
  tenant_inactive: SdkErrorCode.AUTH_ERROR,
  not_found: SdkErrorCode.NETWORK_ERROR,
  device_limit_reached: SdkErrorCode.DEVICE_LIMIT_REACHED,
  invalid_signed_prekey_signature: SdkErrorCode.INVALID_SIGNATURE,
  storage_unavailable: SdkErrorCode.NETWORK_ERROR,
  bad_request: SdkErrorCode.NETWORK_ERROR,
  internal_error: SdkErrorCode.NETWORK_ERROR,
};

File: flutter_sdk/lib/src/errors/sdk_error.dart
/// Sealed error hierarchy for all SDK errors.
///
/// All public Future-returning methods throw subtypes of [SdkError].
/// No raw exceptions from dependencies are surfaced to callers.
///
/// Example:
/// ```dart
/// try {
///   final convo = await client.openConversation('user-123');
/// } on AuthError catch (e) {
///   print('Auth failed: ${e.message}');
/// } on NetworkError catch (e) {
///   print('Network error ${e.statusCode}: ${e.message}');
/// } on SdkError catch (e) {
///   print('SDK error: $e');
/// }
/// ```
sealed class SdkError implements Exception {
  const SdkError();
}

/// A network or HTTP error from the REST transport.
class NetworkError extends SdkError {
  /// HTTP status code, null for connection-level errors.
  final int? statusCode;

  /// Human-readable error message.
  final String message;

  const NetworkError({this.statusCode, required this.message});

  @override
  String toString() => 'NetworkError(statusCode: $statusCode, message: $message)';
}

/// Authentication or authorization failure.
class AuthError extends SdkError {
  final String message;

  const AuthError({required this.message});

  @override
  String toString() => 'AuthError(message: $message)';
}

/// A received message could not be decrypted.
class DecryptionError extends SdkError {
  /// The conversation containing the undecryptable message.
  final String conversationId;

  /// The sequence number of the undecryptable message.
  final int seq;

  const DecryptionError({required this.conversationId, required this.seq});

  @override
  String toString() =>
      'DecryptionError(conversationId: $conversationId, seq: $seq)';
}

/// X3DH key exchange with a recipient failed.
class KeyExchangeError extends SdkError {
  /// The user ID of the intended recipient.
  final String recipientUserId;

  /// Reason for the failure.
  final String reason;

  const KeyExchangeError({required this.recipientUserId, required this.reason});

  @override
  String toString() =>
      'KeyExchangeError(recipientUserId: $recipientUserId, reason: $reason)';
}

/// The session store failed to persist or load state.
class StorageError extends SdkError {
  final String reason;

  const StorageError({required this.reason});

  @override
  String toString() => 'StorageError(reason: $reason)';
}

/// No session was found for the given device ID.
///
/// Thrown when [ChatClientConfig.deviceId] is non-null but no matching
/// key material exists in the [SessionStore].
class SessionNotFoundError extends SdkError {
  final String deviceId;

  const SessionNotFoundError({required this.deviceId});

  @override
  String toString() => 'SessionNotFoundError(deviceId: $deviceId)';
}

/// The file to be uploaded exceeds the 100 MB limit.
class FileTooLargeError extends SdkError {
  final int sizeBytes;

  const FileTooLargeError({required this.sizeBytes});

  @override
  String toString() =>
      'FileTooLargeError(sizeBytes: $sizeBytes, limitBytes: ${100 * 1024 * 1024})';
}

/// An unexpected error not covered by other subtypes.
class UnknownError extends SdkError {
  final Object cause;

  const UnknownError({required this.cause});

  @override
  String toString() => 'UnknownError(cause: $cause)';
}

/// Maps a server [errorCode] and [statusCode] to the appropriate [SdkError].
SdkError sdkErrorFromApiResponse({
  required int statusCode,
  String? errorCode,
  required String message,
}) {
  switch (errorCode) {
    case 'unauthorized':
    case 'forbidden':
    case 'unknown_tenant':
    case 'tenant_inactive':
      return AuthError(message: message);
    case 'invalid_signed_prekey_signature':
      return KeyExchangeError(recipientUserId: '', reason: message);
    case 'storage_unavailable':
      return NetworkError(statusCode: 503, message: message);
    case 'not_found':
      return NetworkError(statusCode: 404, message: message);
    case 'device_limit_reached':
      return NetworkError(statusCode: 409, message: message);
    case 'bad_request':
      return NetworkError(statusCode: 400, message: message);
    case 'internal_error':
      return NetworkError(statusCode: 500, message: message);
    default:
      if (statusCode == 401 || statusCode == 403) {
        return AuthError(message: message);
      }
      return NetworkError(statusCode: statusCode, message: message);
  }
}
